[scroll-anchoring] Exclude isolated subtrees

nmoucht opened the pull request:
#### cf327c0480206b1724eba8ddf418bafadf7aa5c6
<pre>
[scroll-anchoring] Exclude isolated subtrees
<a href="https://bugs.webkit.org/show_bug.cgi?id=262335">https://bugs.webkit.org/show_bug.cgi?id=262335</a>
<a href="https://rdar.apple.com/116205746">rdar://116205746</a>

Reviewed by Simon Fraser.

We should exclude subtrees that are marked as contain: paint and are
offscreen, as any descendents would be clipped to the parent. This doesn&apos;t
quite fix contain-paint-offscreen-container.html but we do select the correct
anchor element after this change.

* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::ScrollAnchoringController::examineCandidate):

Canonical link: <a href="https://commits.webkit.org/270794@main">https://commits.webkit.org/270794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9784323cb91e1386ffb9a7d492414285cacab86e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27650 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24153 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26725 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6802 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2430 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24146 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26648 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3855 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22694 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3412 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3437 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23670 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29087 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24069 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29724 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1687 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27618 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4889 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3944 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3412 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3788 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -188,21 +188,22 @@ CandidateExaminationResult ScrollAnchoringController::examineCandidate(Element&
+        auto boxRect = renderer->absoluteBoundingBoxRect();
+        bool intersects = containingRect.intersects(boxRect);
+
+        if (renderer->style().overflowAnchor() == OverflowAnchor::None || renderer->isStickilyPositioned() || renderer->isFixedPositioned() || renderer->isPseudoElement() || renderer->isAnonymousBlock() || (renderer->isAbsolutelyPositioned() && absolutePositionedElementOutsideScroller(*renderer, m_owningScrollableArea)) || (!intersects && renderer->style().containsPaint()))
```
I don't see any mention of `contain: paint` in the spec (https://www.w3.org/TR/css-scroll-anchoring-1/).

smfr replied:
I think this is just an optimization when you know it's not intersecting.

This line is getting really long. I suggest you write it using a lambda.

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-anchoring] Exclude isolated subtrees
https://bugs.webkit.org/show_bug.cgi?id=262335
rdar://116205746

Reviewed by Simon Fraser.

We should exclude subtrees that are marked as contain: paint and are
offscreen, as any descendents would be clipped to the parent. This doesn't
quite fix contain-paint-offscreen-container.html but we do select the correct
anchor element after this change.

* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::ScrollAnchoringController::examineCandidate):

Canonical link: https://commits.webkit.org/270794@main
